plugins/thunderbolt-power/fu-plugin-thunderbolt-power.c

```diff
@@ -0,0 +1,272 @@
+/* -*- mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
+ *
+ * Copyright (C) 2017 Dell Inc.
+ *
+ * Licensed under the GNU General Public License Version 2
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "config.h"
+
+#include <errno.h>
+#include <fcntl.h>
+#include <gio/gio.h>
+#include <glib.h>
+#include <glib/gstdio.h>
+#include <gudev/gudev.h>
+
+#include "fu-plugin-vfuncs.h"
+#include "fu-device-metadata.h"
+
+#ifndef HAVE_GUDEV_232
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(GUdevDevice, g_object_unref)
+#endif
+
+/* empirically mesaured amount of time for the TBT device to come and go */
+#define TBT_NEW_DEVICE_TIMEOUT	2
+
+struct FuPluginData {
+	GUdevClient   *udev;
+	gchar         *force_path;
+	gboolean       needs_forcepower;
+};
+
+static void
+fu_plugin_thunderbolt_power_get_path (FuPlugin *plugin)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	g_autoptr(GList) devices = NULL;
+	const gchar *basepath;
+	const gchar *driver;
+
+	devices = g_udev_client_query_by_subsystem (data->udev, "wmi");
+	for (GList* l = devices; l != NULL; l = l->next) {
+		g_autofree gchar *built_path = NULL;
+		g_autoptr(GUdevDevice) device = l->data;
+
+		/* only supports intel-wmi-thunderbolt for now */
+		driver = g_udev_device_get_driver (device);
+		if (g_strcmp0 (driver, "intel-wmi-thunderbolt") != 0)
+			continue;
+
+		/* check for the attribute to be loaded */
+		basepath = g_udev_device_get_sysfs_path (device);
+		if (basepath == NULL)
+			continue;
+		built_path = g_build_path ("/", basepath,
+					   "force_power", NULL);
+		if (g_file_test (built_path, G_FILE_TEST_IS_REGULAR)) {
+			data->force_path = g_steal_pointer (&built_path);
+			return;
+		}
+	}
+
+	/* driver went away */
+	if (data->force_path)
```

Comment:
I'm not sure I get it.
The only place this is set is line 72 above and then we don't reach this.
If this is from a previous run of this method, then line 72 need to `g_free` `force_path` before assigning to it.

Reply:
There's two ways to enter the function:
1. plugin initialization
2.  intel-wmi-thunderbolt module change event

If you initialize with the module loaded this gets set during init, and then you unload the module it needs to be unset when it's called again.  If the module gets loaded again, you would have another change event and it would get set again.

If you initialize without the module loaded this hasn't yet been set.  You load the module and it gets set.  If you unload again it needs to get unset.

Reply:
I still think it's better to move this before the loop, otherwise, if the module generates change event for any other reason (or got removed and reloaded before this had a chance to run) there is still a leak.

Reply:
That's true.  I'll move it.